core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -0,0 +1,193 @@
+package cats
+package data
+
+import cats.instances.list._
+import cats.syntax.eq._
+
+import scala.annotation.tailrec
+import scala.collection.mutable.ListBuffer
+
+/**
+ * A data type which represents a non empty list of A, with
+ * single element (head) and optional structure (tail).
+ */
+final case class NonEmptyList[A](head: A, tail: List[A]) {
+
+  /**
+   * Return the head and tail into a single list
+   */
+  def toList: List[A] = head :: tail
+
+  /**
+   *  Applies f to all the elements of the structure
+   */
+  def map[B](f: A => B): NonEmptyList[B] =
+    NonEmptyList(f(head), tail.map(f))
+
+  def ++(l: List[A]): NonEmptyList[A] =
+    NonEmptyList(head, tail ++ l)
+
+  def flatMap[B](f: A => NonEmptyList[B]): NonEmptyList[B] =
+    f(head) ++ tail.flatMap(f andThen (_.toList))
+
+  /**
+   * remove elements not matching the predicate
+   */
+  def filter(p: A => Boolean): List[A] =
+    toList.filter(p)
+
+  /**
+   * Append another NonEmptyList
+   */
+  def concat(other: NonEmptyList[A]): NonEmptyList[A] =
+    NonEmptyList(head, tail ::: other.toList)
+
+  /**
+   * Find the first element matching the predicate, if one exists
+   */
+  def find(p: A => Boolean): Option[A] =
+    toList.find(p)
+
+  /**
+   * Check whether at least one element satisfies the predicate
+   */
+  def exists(p: A => Boolean): Boolean =
+    p(head) || tail.exists(p)
+
+  /**
+   * Check whether all elements satisfy the predicate
+   */
+  def forall(p: A => Boolean): Boolean =
+    p(head) && tail.forall(p)
+
+  /**
+   * Left-associative fold on the structure using f.
+   */
+  def foldLeft[B](b: B)(f: (B, A) => B): B =
+    tail.foldLeft(f(b, head))(f)
+
+  /**
+   * Right-associative fold on the structure using f.
+   */
+  def foldRight[B](lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B] =
+    Foldable[List].foldRight(toList, lb)(f)
+
+  /**
+   * Left-associative reduce using f.
+   */
+  def reduceLeft(f: (A, A) => A): A =
+    tail.foldLeft(head)(f)
+
+  def coflatMap[B](f: NonEmptyList[A] => B): NonEmptyList[B] = {
+    @tailrec def consume(as: List[A], buf: ListBuffer[B]): List[B] =
```

Comment:
why pass the mutable `buf`? Can we capture to make more clear there is only one:
```scala
val buf = ListBuffer.empty
@tailrec def consume(as: List[A]): List[B] = {
...
}
```